merge code for llvm.emit_c_interface into convertFuncOpToLLVMFuncOp

fengxie opened the pull request:
In today's repo, attribute `llvm.emit_c_interface` of func op is handled outside of `mlir::convertFuncOpToLLVMFuncOp` in `FuncOpConversion` pattern. In some cases, `FuncOpConversion` can't be directly re-used, but we still want to re-use the code to emit c interface for `llvm.emit_c_interface`.

Changes in this PR
* move the code to generate c with "llvm.emit_c_interface" interface into `mlir::convertFuncOpToLLVMFuncOp` to be able to re-use it.
* added unit test to verify c interface for jit can be generated correctly if only call `convertFuncOpToLLVMFuncOp`.
* removed `FuncOpConversionBase`

fengxie committed:
merge code for llvm.emit_c_interface into convertFuncOpToLLVMFuncOp

fengxie committed:
fixed comments and remove un-used headers

joker-eph commented:
The change LGTM but I'd want to ask @ftynse : do you see a reason why the pattern included this code and not the "utility API" before?

fengxie committed:
updated test to only test mlir::convertFuncOpToLLVMFuncOp and remove FuncOpConversionBase

fengxie committed:
delete unused include

fengxie commented:
Updated test to only test utility function `convertFuncOpToLLVMFuncOp` which is main purpose of this change.

grypp commented on mlir/test/lib/Conversion/FuncToLLVM/TestConvertFuncOp.cpp:
```diff
@@ -0,0 +1,83 @@
+    registry.insert<LLVM::LLVMDialect>();
+  }
+
+  [[nodiscard]] StringRef getArgument() const final {
```
I haven't seen `[[nodiscard]]` usage. do we need it?

fengxie replied:
clang-tidy suggested to add.

ftynse replied:
It may be an overly eager configuration, we usually don't mark these methods nodiscard, so let's not do that here.

fengxie replied:
`[[nodiscard]]` is removed.

ftynse commented on mlir/test/lib/Conversion/FuncToLLVM/TestConvertFuncOp.cpp:
```diff
@@ -0,0 +1,83 @@
+    RewritePatternSet patterns(ctx);
+    patterns.add<FuncOpConversion>(ctx);
+
+    populateFuncToLLVMConversionPatterns(typeConverter, patterns, nullptr);
```
Wouldn't this add the regular func conversion pattern in addition to the one added above? So we don't actually know what we are testing?

fengxie replied:
Thanks a lot for suggestion. I actually started with RewriterBase &. But RewriterBase doesn't have inlineRegionBefore and convertRegionTypes. I prefer RewriterBase & if it works. Do you have suggestion on it?

```c++
  rewriter.inlineRegionBefore(funcOp.getFunctionBody(), newFuncOp.getBody(),
                              newFuncOp.end());
  if (failed(rewriter.convertRegionTypes(&newFuncOp.getBody(), converter,
                                         &result))) {
    return rewriter.notifyMatchFailure(funcOp,
                                       "region types conversion failed");
  }
```

fengxie replied:
> Wouldn't this add the regular func conversion pattern in addition to the one added above? So we don't actually know what we are testing?

Yes, this adds regular func conversion patterns. I created a test only have "llvm.emit_c_interface" to only test it is handled by this function.

ftynse replied:
I don't see why we need the regular func conversion pattern here. It should be enough to exercise the new pattern, and we want tests to be as minimal as possible.

fengxie replied:
Ah, thanks for explanation. I think I misunderstood your feedback. What I need to do is to delete this line right?

I just deleted this line and replaced it with pattern conversion for return op to llvm ( I need this to convert full func op ). Please let me know if this addressed the comment. Thanks in advance!

joker-eph replied:
Technically we don't even need to use a pattern, just a walk() should do right?

fengxie replied:
For this one, I have to use pattern as `convertFuncOpToLLVMFuncOp` requires pattern rewriter which I can't create outside of pattern. For `ReturnOpConversion`, I choose to use pattern as I already have pattern rewriter setup in place for funcOp.

```c++
patterns.add<FuncOpConversion>(ctx);
```

fengxie replied:
@ftynse if got a chance, can you review latest changes are good to address feedbacks?
* deleting general pattern and replace it with return op -> llvm.return pattern.

```c++
    populateFuncToLLVMConversionPatterns(typeConverter, patterns, nullptr);
```

ftynse reviewed (changes requested):
Thanks for the patch! I suppose this is the case of "nobody needed this yet".

Can we do one step further and actually generalize this function to be usable without conversion patterns? It can take a `RewriterBase &` instead of `ConversionPatternRewriter` as far as I can see. Then it becomes usable standalone and the test can be simplified to actually test the function and not the function + conversion infra.

fengxie commented:
> Thanks for the patch! I suppose this is the case of "nobody needed this yet".
> 
> Can we do one step further and actually generalize this function to be usable without conversion patterns? It can take a `RewriterBase &` instead of `ConversionPatternRewriter` as far as I can see. Then it becomes usable standalone and the test can be simplified to actually test the function and not the function + conversion infra.

I actually started with `RewriterBase &`. But `RewriterBase` doesn't have `inlineRegionBefore` and `convertRegionTypes`. I prefer `RewriterBase &` if it works. Do you have suggestion on it?

```c++
  rewriter.inlineRegionBefore(funcOp.getFunctionBody(), newFuncOp.getBody(),
                              newFuncOp.end());
  if (failed(rewriter.convertRegionTypes(&newFuncOp.getBody(), converter,
                                         &result))) {
    return rewriter.notifyMatchFailure(funcOp,
                                       "region types conversion failed");
  }
```

ftynse commented:
> I actually started with RewriterBase &. But RewriterBase doesn't have inlineRegionBefore and convertRegionTypes. I prefer RewriterBase & if it works. Do you have suggestion on it?

Thanks for checking. In this case, it looks easier to just keep the `ConversionPatternRewriter`. The alternative would be to pass in a callback, but that's too much fuss. Please address the comments about the test and this will be good to go.

fengxie committed:
remove nodiscard

fengxie commented:
@ftynse please let me know if all comments are addressed

* `[[nodiscard]]` was removed
* Regarding this https://github.com/llvm/llvm-project/pull/92986#discussion_r1609599101, does reply here addressed the comment? https://github.com/llvm/llvm-project/pull/92986#discussion_r1609959512

joker-eph reviewed (approved):
LG, but let's wait for @ftynse !

fengxie committed:
remove populateFuncToLLVMConversionPatterns

ftynse reviewed (approved):
Thanks!
